proxy/proxy.go

```diff
@@ -17,8 +17,8 @@ import (
 
 type (
 	RespMiddleware func(resp *http.Response, config *HostConfig, body []byte) ([]byte, error)
-	ReqMiddleware  func(req *http.Request, config *HostConfig, body []byte) ([]byte, error)
-	HostMapper     func(ctx context.Context, r *http.Request) (*HostConfig, error)
+	ReqMiddleware  func(req *httputil.ProxyRequest, config *HostConfig, body []byte) ([]byte, error)
+	HostMapper     func(ctx context.Context, r *http.Request) (context.Context, *HostConfig, error)
```

Comment:
It's no longer possible to use the context properly because we don't know whether we get the context from `req.In` or `req.Out` now. So context needs to be explicit now.